osu.Game.Rulesets.Taiko/UI/TaikoRulesetContainer.cs

```diff
@@ -15,21 +18,45 @@
 using osu.Game.Rulesets.UI;
 using osu.Game.Rulesets.Taiko.Replays;
 using OpenTK;
+using OpenTK.Input;
 using System.Linq;
 using osu.Framework.Input;
+using System.Collections.Generic;
 
 namespace osu.Game.Rulesets.Taiko.UI
 {
     public class TaikoRulesetContainer : ScrollingRulesetContainer<TaikoPlayfield, TaikoHitObject>
     {
+        private readonly HashSet<Key> centreKeys = new HashSet<Key>();
+        private readonly HashSet<Key> rimKeys = new HashSet<Key>();
+        private AudioManager audio;
+        private IEnumerable<KeyBinding> keyBindings;
+
         public TaikoRulesetContainer(Ruleset ruleset, WorkingBeatmap beatmap, bool isForCurrentRuleset)
             : base(ruleset, beatmap, isForCurrentRuleset)
         {
         }
 
         [BackgroundDependencyLoader]
-        private void load()
+        private void load(AudioManager audio, KeyBindingStore store)
         {
+            keyBindings = store.Query(Ruleset.RulesetInfo.ID, Ruleset.AvailableVariants?.First() ?? 0).Cast<KeyBinding>();
+            if (keyBindings.Count() == 0)
+                keyBindings = Ruleset.GetDefaultKeyBindings();
+
+            foreach (var kb in keyBindings)
+            {
+                var key = (Key)(kb.KeyCombination.Keys as InputKey[]).First();
+                var action = kb.GetAction<TaikoAction>();
+
+                if (action == TaikoAction.LeftCentre || action == TaikoAction.RightCentre)
+                    centreKeys.Add(key);
+
+                if (action == TaikoAction.LeftRim || action == TaikoAction.RightRim)
+                    rimKeys.Add(key);
+            }

```

Comment:
Instead of this, make `TaikoPlayfield` implement `IKeyBindingHandler<TaikoAction>`.

Then you'll have two cases: Either a hitobject has been hit, in which case it will block input from propagating outwards, or a hitobject hasn't been hit and the input propagates outwards.

* In the case that the input does not propagate outwards, the hitobject handles playing of its sample.
* In the case that input does propagate outwards, `TaikoPlayfield` will eventually receive an `OnPressed`. This is where the `ControlPoint`'s samples should be played.

This way, you don't need all of this or the `PlaysSamples` stuff inside DrawableHitObject.